core/translate/main_loop.rs

```diff
@@ -1141,44 +1140,54 @@ fn emit_seek(
     seek_cursor_id: usize,
     start_reg: usize,
     loop_end: BranchOffset,
-    is_index: bool,
+    seek_index: Option<&Arc<Index>>,
 ) -> Result<()> {
+    let is_index = seek_index.is_some();
     let Some(seek) = seek_def.seek.as_ref() else {
         // If there is no seek key, we start from the first or last row of the index,
         // depending on the iteration direction.
+        //
+        // Also, if we will encounter NULLs in the index at the beginning of iteration (Forward + Asc OR Backward + Desc)
+        // then, we must explicitly skip them as seek always has some bound condition over indexed column (e.g. c < ?, c >= ?, ...)
+        //
+        // note that table seek has some rules to convert seek key values to the integer affinity + it has some logic to check for explicit NULL searches
+        // so, we emit simple Rewind/Last in case of search over table BTree
+        // (this is safe as table BTree keys are always non-null single integer)
         match seek_def.iter_dir {
             IterationDirection::Forwards => {
-                // table seek has some rules to convert seek key values to the integer affinity + it has some logic to check for explicit NULL searches
-                // so, we emit simple Rewind in case of search over table BTree
-                // (this is safe as table BTree keys are always non-null single integer)
-                if !is_index {
-                    program.emit_insn(Insn::Rewind {
+                if seek_index.is_some() && seek_index.unwrap().columns[0].order == SortOrder::Asc {
+                    program.emit_null(start_reg, None);
+                    program.emit_insn(Insn::SeekGT {
+                        is_index,
                         cursor_id: seek_cursor_id,
-                        pc_if_empty: loop_end,
+                        start_reg,
+                        num_regs: 1,
+                        target_pc: loop_end,
                     });
                 } else {
-                    // the seek always has some bound condition over indexed column (e.g. c < ?, c >= ?, ...)
-                    // so, NULL always must be filtered out
-                    // (note, that complex filters like c < ? OR c IS NULL will produce Scan operation instead of Search)
-                    program.emit_insn(Insn::Null {
-                        dest: start_reg,
-                        dest_end: None,
+                    program.emit_insn(Insn::Rewind {
+                        cursor_id: seek_cursor_id,
+                        pc_if_empty: loop_end,
                     });
-                    program.emit_insn(Insn::SeekGT {
+                }
+            }
+            IterationDirection::Backwards => {
+                if seek_index.is_some() && seek_index.unwrap().columns[0].order == SortOrder::Desc {
```

Comment:
no need to fix this here since it's just a cosmetic convenience thing, but `seek_index.is_some_and(|unwrapped_value| unwrapped_value.columns..........` is a nice pattern